### PR TITLE
ENH: add support for NEP 35 array constructors (`empty`, `ones`, `zeros` and `full`) with `like=Quantity(...)`)

### LINKED ISSUE #17001: Investigate enabling array creation routines for quantities
### What is the problem this feature will solve?

Currently, [Numpy array creation functions cannot be used to initialize Quantity](https://docs.astropy.org/en/stable/known_issues.html#id4). However, @maxnoe speculates:

> I think it could be implemented now using the like option. Passing an empty quantity to like now creates an error inside astropy:
> 
> ```
> In [12]: np.arange(0 * u.m, 0.1 * u.m, 0.01 * u.m, like=u.Quantity([], u.m))
> WARNING: function 'arange' is not known to astropy's Quantity. Will run it anyway, hoping it will treat ndarray subclasses correctly. Please raise an issue at https://github.com/astropy/astropy/issues. [astropy.units.quantity]
> ---------------------------------------------------------------------------
> AttributeError                            Traceback (most recent call last)
> Cell In[12], line 1
> ----> 1 np.arange(0 * u.m, 0.1 * u.m, 0.01 * u.m, like=u.Quantity([], u.m))
> 
> File ~/.local/conda/envs/cta-dev/lib/python3.11/site-packages/astropy/units/quantity.py:1888, in Quantity.__array_function__(self, function, types, args, kwargs)
>    1880 else:
>    1881     warnings.warn(
>    1882         f"function '{function.__name__}' is not known to astropy's Quantity."
>    1883         " Will run it anyway, hoping it will treat ndarray subclasses"
>    (...)
>    1886         AstropyWarning,
>    1887     )
> -> 1888     return super().__array_function__(function, types, args, kwargs)
>    1890 # If unit is None, a plain array is expected (e.g., boolean), which
>    1891 # means we're done.
>    1892 # We're also done if the result was NotImplemented, which can happen
>    1893 # if other inputs/outputs override __array_function__;
>    1894 # hopefully, they can then deal with us.
>    1895 if unit is None or result is NotImplemented:
> 
> AttributeError: 'builtin_function_or_method' object has no attribute '_implementation'
> ```

### Describe the desired outcome

Enable the Numpy array creation routines for use with quantities: [full](https://numpy.org/doc/stable/reference/generated/numpy.full.html#numpy.full), [zeros](https://numpy.org/doc/stable/reference/generated/numpy.zeros.html#numpy.zeros), [ones](https://numpy.org/doc/stable/reference/generated/numpy.ones.html#numpy.ones), [empty](https://numpy.org/doc/stable/reference/generated/numpy.empty.html#numpy.empty), and [arange](https://numpy.org/doc/stable/reference/generated/numpy.arange.html#numpy.arange).

### Additional context

CC @neutrinoceros.

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Follow up to #17059, fixes #17001
Note that this is not a complete implementation of NEP 35 functions, but it does cover everything that was listed on #17001. I plan to open another issue to track what's left to do.
While working on these functions, I found that the mechanism suggested by @mhvk in #17059 for defaulting the output unit to that of the `like` argument made most sense for functions that don't take other arguments for which a quantity is meaningful, so I implemented it here and retro-fitted it in my implementation for `np.arange` too, with the particularity that I'm passing it under a different, private name (`_like`) to better signal that this argument is internal and not passed down by numpy itself.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
I'll let units maintainers to decide if this needs a What's New. Thanks!

File: astropy/units/tests/test_quantity_non_ufuncs.py
Comment (mhvk) [inline astropy/units/tests/test_quantity_non_ufuncs.py:2716]: Can we add a comment referring to your numpy issue? Hopefully, we won't need to use `SUPPORTED_NEP35_FUNCTIONS` here eventually.

File: astropy/units/tests/test_quantity_non_ufuncs.py
Comment (mhvk) [inline astropy/units/tests/test_quantity_non_ufuncs.py:2839]: This could go with my suggestion.

**Review by mhvk — COMMENTED** [review]
Looks good, but with a suggestion.  Also, how about `asarray`?

**Comment by mhvk** [comment]
> I'll let units maintainers to decide if this needs a What's New. Thanks!

A what's new makes sense, if we support all of them!

**Comment by neutrinoceros** [comment]
> A what's new makes sense, if we support all of them!

I can add this as a todo to the global issue I'll write tomorrow. Would that work ?

**Comment by neutrinoceros** [comment]
> how about `asarray`?

I'm focusing on what was requested in #17001 to avoid bloating this PR for now, and `asarray` is part of a subset of functions that I think belong together (`array`, `asarray`, `asanyarray`, `ascontiguousarray` and `asfortranarray`) and plan to tackle next.

File: astropy/units/tests/test_quantity_non_ufuncs.py
Comment (neutrinoceros) [reply astropy/units/tests/test_quantity_non_ufuncs.py:2841]: not completely though: since re-exposing `like` doesn't really make sense, I'm still keeping this branch (simplified, though)

File: astropy/units/quantity.py
Comment (neutrinoceros) [reply astropy/units/quantity.py]: Done. I was confused at first by the use of `True` as a sentinel value, which I believe is sufficient proof that it's not very clear, so I added my own, but the logic is yours !

**Comment by neutrinoceros** [comment]
Alright, I implemented @mhvk's suggestion, so this should be ready to go. I'll also write the follow up issue now.

**Comment by neutrinoceros** [comment]
I almost forgot to also update documentation, but it's done now

File: astropy/units/quantity_helper/function_helpers.py
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py:495]: Thinking about this, I think you still need `unit_from_first`, otherwise if `stop` has no unit and `start` has one, that unit from `start` will be taken but discarded here. I think it may be best to have an explicit test that if `stop` has no unit, all other values also have no unit. That probably can then remain a separate branch, with this piece only happening for the case that `stop` has a unit.

File: astropy/units/quantity_helper/function_helpers.py
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py:446]: Why the camel case? I thought that was meant more for classes. Just `unit_from_like` or `UNIT_FROM_LIKE` (no real reason for it to be private, I think; none of these are exported up).

File: astropy/units/quantity_helper/function_helpers.py
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py:513]: This implementation function is super short and still has to distinguish between numpy versions - just put them directly in the functions above?

File: astropy/units/quantity_helper/function_helpers.py
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py:530]: Same as above. Alternatively, one could have an implementation for all those creation functions (i.e., also pass a `function=np.ones` along).

File: astropy/units/quantity_helper/function_helpers.py
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py:538]: Does `zeros` not have the `device` argument?

File: astropy/units/quantity_helper/function_helpers.py
Comment (mhvk) [inline astropy/units/quantity_helper/function_helpers.py:550]: Here one could still call the overall implementation, and then get the unit after.

**Review by mhvk — COMMENTED** [review]
Thanks! One larger worry about the use of the unit in `arange`, otherwise small bits & pieces.

File: astropy/units/quantity_helper/function_helpers.py
Comment (neutrinoceros) [reply astropy/units/quantity_helper/function_helpers.py:538]: It doesn't. Otherwise we'd be seeing at least one test fail, but I admit I too was surprised that its signature differs from that of ones. It's probably either an oversight or reflects that ones is part of the Array API and zeros isn't.

**Comment by mhvk** [comment]
@neutrinoceros - I made a quick PR to yours with an implementation using `function_helper` - which is really the one that is supposed to deal with cases where functions are very similar. See https://github.com/neutrinoceros/astropy/pull/25

File: astropy/units/quantity_helper/function_helpers.py
Comment (neutrinoceros) [reply astropy/units/quantity_helper/function_helpers.py:446]: > Why the camel case?

I don't know 😅

File: astropy/units/quantity_helper/function_helpers.py
Comment (neutrinoceros) [reply astropy/units/quantity_helper/function_helpers.py]: I don't feel comfortable generalizing early as I worry we could get trapped when later expanding the list of implemented function, so I'd like to avoid designing a general implementation for now if that's okay.

**Comment by neutrinoceros** [comment]
@mhvk here's my attempt to reply to your latest review, but there may still be a couple points to discuss.

**Comment by mhvk** [comment]
@neutrinoceros - I rebased my PR to your PR

**Comment by neutrinoceros** [comment]
aaaand merged. Thanks !

File: astropy/units/tests/test_quantity_non_ufuncs.py
Comment (mhvk) [inline astropy/units/tests/test_quantity_non_ufuncs.py:468]: You could factor out the test by making a new class
```
class TestArrayCreation(BasicTestSetup):
    def check(self, func, *args, **kwargs):
        o = func(*args, **kwargs, like=self.q)
        assert type(o) is type(self.q)
        assert o.unit == self.q.unit
        expected = func(*args, **kwargs) << self.q.unit
        assert_array_equal(o, expected)
```

Then only for `full` do you need to keep the test with `fill_value` with a unit.

**Review by mhvk — APPROVED** [review]
All looks good now, but one suggestion still in-line (though fine to merge without it).

**Comment by neutrinoceros** [comment]
If that's okay I'd prefer to apply your suggestion in a follow up PR, so I can rebase #17025 on main sooner rather than later and avoid stacking any further. Thanks !

**Comment by mhvk** [comment]
OK, let's get this one in!

File: astropy/units/tests/test_quantity_non_ufuncs.py
Comment (neutrinoceros) [reply astropy/units/tests/test_quantity_non_ufuncs.py:468]: for reference, I've integrated this suggestion into #17125

**Pull request merged**
